src/server/pmix_server_get.c

```diff
@@ -0,0 +1,551 @@
+        pmix_rank_info_t *info;
+        bool found = false;
+
+        if (0 != strncmp(nptr->nspace, cd->proc.nspace, PMIX_MAX_NSLEN) ) {
+            continue;
+        }
+
+        PMIX_LIST_FOREACH(info, &nptr->server->ranks, pmix_rank_info_t) {
+            if (info->rank == cd->proc.rank) {
+                found = true;  // we will satisy this request upon commit from new proc
+                break;
+            }
+        }
+
+        /* if not found - this is remote process and we need to send
+         * corresponding direct modex request */
+        if( !found ){
+            if( NULL != pmix_host_server.direct_modex ){
+                pmix_host_server.direct_modex(&cd->proc, cd->info, cd->ninfo, dmdx_cbfunc, cd);
+            } else {
+                pmix_dmdx_request_t *req, *req_next;
+                PMIX_LIST_FOREACH_SAFE(req, req_next, &cd->loc_reqs, pmix_dmdx_request_t) {
+                    req->cbfunc(PMIX_ERR_NOT_FOUND, NULL, 0, req->cbdata, NULL, NULL);
+                    pmix_list_remove_item(&cd->loc_reqs, &req->super);
+                    PMIX_RELEASE(req);
+                }
+                pmix_list_remove_item(&pmix_server_globals.local_reqs, &cd->super);
+                PMIX_RELEASE(cd);
+            }
+        }
+    }
+}
+
+static pmix_status_t _satisfy_request(pmix_hash_table_t *ht, int rank,
+                                      pmix_modex_cbfunc_t cbfunc, void *cbdata)
+{
+    pmix_status_t rc;
+    pmix_value_t *val;
+    char *data;
+    size_t sz;
+    pmix_buffer_t xfer, pbkt, *xptr;
+
+    /* check to see if this data already has been
+     * obtained as a result of a prior direct modex request from
+     * a remote peer, or due to data from a local client
+     * having been committed */
+    rc = pmix_hash_fetch(ht, rank, "modex", &val);
+    if (PMIX_SUCCESS == rc && NULL != val) {
+        /* the client is expecting this to arrive as a byte object
+         * containing a buffer, so package it accordingly */
+        PMIX_CONSTRUCT(&pbkt, pmix_buffer_t);
+        PMIX_CONSTRUCT(&xfer, pmix_buffer_t);
+        xptr = &xfer;
+        PMIX_LOAD_BUFFER(&xfer, val->data.bo.bytes, val->data.bo.size);
+        pmix_bfrop.pack(&pbkt, &xptr, 1, PMIX_BUFFER);
+        xfer.base_ptr = NULL; // protect the passed data
+        xfer.bytes_used = 0;
+        PMIX_DESTRUCT(&xfer);
+        PMIX_UNLOAD_BUFFER(&pbkt, data, sz);
+        PMIX_DESTRUCT(&pbkt);
+        PMIX_VALUE_RELEASE(val);
+        /* pass it back */
+        cbfunc(rc, data, sz, cbdata, relfn, data);
+        return rc;
+    }
+    return PMIX_ERR_NOT_FOUND;
+}
+
+/* Resolve pending requests to this namespace/rank */
+pmix_status_t pmix_pending_resolve(pmix_nspace_t *nptr, int rank,
+                                   pmix_status_t status, pmix_dmdx_local_t *lcd)
+{
+    pmix_dmdx_local_t *cd;
+
+    /* find corresponding request (if exists) */
+    if (NULL == lcd && NULL != nptr) {
```

Comment:
I suggest to put
if( NULL == nptr && NULL == lcd ){
    return PMIX_SUCCESS;
}else if( NULL == nptr ){
   nptr = lcd.proc.nspace;
}

at the beginning of the function and forget about nptr == NULL in the rest of the code.